File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/PeerObserver.kt
package com.cloudwebrtc.webrtc.proxy

import android.os.Handler
import android.os.Looper
import com.cloudwebrtc.webrtc.model.IceCandidate
import com.cloudwebrtc.webrtc.model.IceConnectionState
import com.cloudwebrtc.webrtc.model.IceGatheringState
import com.cloudwebrtc.webrtc.model.SignalingState
import org.webrtc.*
import org.webrtc.IceCandidate as WIceCandidate

/**
 * Implementor of a [PeerConnection.Observer] notifying a [PeerConnectionProxy]
 * about [PeerConnection] events.
 */
class PeerObserver : PeerConnection.Observer {
    /**
     * [PeerConnectionProxy] being notified about all events.
     */
    private var peer: PeerConnectionProxy? = null

    override fun onSignalingChange(signallingState: PeerConnection.SignalingState?) {
        if (signallingState != null) {
            Handler(Looper.getMainLooper()).post {
                peer?.observableEventBroadcaster()
                    ?.onSignalingStateChange(
                        SignalingState.fromWebRtc(
                                signallingState
                        )
                    )
            }
        }
    }

    override fun onIceConnectionChange(iceConnectionState: PeerConnection.IceConnectionState?) {
        if (iceConnectionState != null) {
            Handler(Looper.getMainLooper()).post {
                peer?.observableEventBroadcaster()
                    ?.onIceConnectionStateChange(
                        IceConnectionState.fromWebRtc(
                                iceConnectionState
                        )
                    )
            }
        }
    }

    override fun onIceGatheringChange(iceGatheringState: PeerConnection.IceGatheringState?) {
        if (iceGatheringState != null) {
            Handler(Looper.getMainLooper()).post {
                peer?.observableEventBroadcaster()
                    ?.onIceGatheringStateChange(
                        IceGatheringState.fromWebRtc(
                                iceGatheringState
                        )
                    )
            }
        }
    }

    override fun onIceCandidate(candidate: WIceCandidate?) {
        if (candidate != null) {
            Handler(Looper.getMainLooper()).post {
                peer?.observableEventBroadcaster()
                    ?.onIceCandidate(IceCandidate.fromWebRtc(candidate))
            }
        }
    }

    override fun onTrack(transceiver: RtpTransceiver?) {
        if (transceiver != null) {
            Handler(Looper.getMainLooper()).post {
                val receiver = transceiver.receiver
                val track = receiver.track()
                if (track != null) {
                    val transceivers = peer?.getTransceivers()!!
                    for (trans in transceivers) {
                        if (trans.getReceiver().id() == receiver.id()) {
                            peer?.observableEventBroadcaster()
                                ?.onTrack(MediaStreamTrackProxy(track), trans)
                        }
                    }
                }
            }
        }
        super.onTrack(transceiver)
    }

    override fun onRenegotiationNeeded() {
        Handler(Looper.getMainLooper()).post {
            peer?.observableEventBroadcaster()?.onNegotiationNeeded()
        }
    }

    override fun onIceConnectionReceivingChange(receiving: Boolean) {}
    override fun onIceCandidatesRemoved(candidates: Array<out WIceCandidate>?) {}
    override fun onAddStream(stream: MediaStream?) {}
    override fun onRemoveStream(stream: MediaStream?) {}
    override fun onDataChannel(chan: DataChannel?) {}

    /**
     * Sets the [PeerConnectionProxy] to be notified about all events.
     */
    fun setPeerConnection(newPeer: PeerConnectionProxy) {
        peer = newPeer
    }
}

File: ios/Classes/proxy/PeerObserver.swift
import Dispatch
import WebRTC

/// Observer for a native `RTCPeerConnectionDelegate`.
class PeerObserver: NSObject, RTCPeerConnectionDelegate {
  /// `PeerConnectionProxy` into which callbacks will be provided.
  var peer: PeerConnectionProxy?

  override init() {}

  /// Sets the underlying `PeerConnectionProxy` for this `PeerObserver`.
  func setPeer(peer: PeerConnectionProxy) {
    self.peer = peer
  }

  /// Fires an `onSignalingStateChange` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _: RTCPeerConnection, didChange stateChanged: RTCSignalingState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onSignalingStateChange(
        state: SignalingState.fromWebRtc(state: stateChanged)
      )
    }
  }

  /// Fires an `onIceConnectionStateChange` callback in the
  /// `PeerConnectionProxy`.
  func peerConnection(
    _: RTCPeerConnection, didChange newState: RTCIceConnectionState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onIceConnectionStateChange(
        state: IceConnectionState.fromWebRtc(state: newState)
      )
    }
  }

  /// Fires an `onConnectionStateChange` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _: RTCPeerConnection, didChange newState: RTCPeerConnectionState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onConnectionStateChange(
        state: PeerConnectionState.fromWebRtc(state: newState)
      )
    }
  }

  /// Fires an `onIceGatheringStateChange` callback in the
  /// `PeerConnectionProxy`.
  func peerConnection(
    _: RTCPeerConnection, didChange newState: RTCIceGatheringState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onIceGatheringStateChange(
        state: IceGatheringState.fromWebRtc(state: newState)
      )
    }
  }

  /// Fires an `onIceCandidate` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _: RTCPeerConnection, didGenerate candidate: RTCIceCandidate
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onIceCandidate(
        candidate: IceCandidate(candidate: candidate)
      )
    }
  }

  /// Fires an `onTrack` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _: RTCPeerConnection, didStartReceivingOn transceiver: RTCRtpTransceiver
  ) {
    DispatchQueue.main.async {
      let receiver = transceiver.receiver
      let track = receiver.track
      if track != nil {
        let transceivers = self.peer!.getTransceivers()
        for trans in transceivers {
          if trans.getReceiver().id() == receiver.receiverId {
            self.peer!.broadcastEventObserver().onTrack(
              track: trans.getReceiver().getTrack(),
              transceiver: trans
            )
          }
        }
      }
    }
  }

  /// Does nothing.
  func peerConnection(
    _: RTCPeerConnection, didAdd _: RTCRtpReceiver,
    streams _: [RTCMediaStream]
  ) {}

  /// Does nothing.
  func peerConnection(
    _: RTCPeerConnection, didRemove receiver: RTCRtpReceiver
  ) {
    DispatchQueue.main.async {
      self.peer!.receiverRemoved(endedReceiver: receiver)
    }
  }

  /// Does nothing.
  func peerConnection(_: RTCPeerConnection, didAdd _: RTCMediaStream) {}

  /// Does nothing.
  func peerConnection(_: RTCPeerConnection, didRemove _: RTCMediaStream) {}

  /// Does nothing.
  func peerConnection(_: RTCPeerConnection, didOpen _: RTCDataChannel) {}

  /// Does nothing.
  func peerConnectionShouldNegotiate(_: RTCPeerConnection) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onNegotiationNeeded()
    }
  }

  /// Does nothing.
  func peerConnection(
    _: RTCPeerConnection, didRemove _: [RTCIceCandidate]
  ) {}
}
